stages/authenticator_validate: autoselect last used 2fa device

gergosimonyi opened the pull request:
## Details

Closes #3488

Users who configure multiple 2fa devices will now land on their last used device's prompt, skipping the device picker. (They can of course go to the device picker from there.)

I've made an exception for static recovery codes as I believe that would cause more confusion than good.

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

If applicable

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make website`)

gergosimonyi committed:
authenticator_validate: autoselect last used device class

gergosimonyi committed:
improve usability of `AuthenticatorValidationStage`

gergosimonyi committed:
don't automatically offer the recovery key authenticator validation

I believe this could confuse users more than help them

kensternberg-authentik commented on web/src/flow/stages/authenticator_validate/AuthenticatorValidateStage.ts:
```diff
@@ -231,22 +234,40 @@ export class AuthenticatorValidateStage
+            lastUsedChallenge.deviceClass !== DeviceClassesEnum.Static &&
+            !this._initialized
+        ) {
+            this._initialized = true;
```
Ah, don't change `property()` or `state()` fields inside `render()`. 

The problem is that I don't quite understand what this code is for.  If you need to set `this.selectedDeviceChallenge`, do it in the `willUpdate()` phase of the render.  [Changes inside `willUpdate()` cannot trigger rerenders.](https://lit.dev/docs/components/lifecycle/#willupdate)

kensternberg-authentik replied:
Also, is there a reason you need to track the `this._initialized`  at all?  What happens if you don't have it?  Could you have the `selectedDeviceChallenge` itself be the sigil you need to know which one to highlight?

gergosimonyi replied:
Let's say the user's last used device is TOTP. Without `this._initialized`, the following could happen:

1. User arrives to Authenticator Validate stage.
2. TOTP is automatically selected and the user is presented with the TOTP validation screen.
3. User decides to use webauthn instead, so clicks "Return to device picker".
4. TOTP is automatically selected again.

Effectively, without some countermeasure, if a device is autoselected the user can't go back to the device picker. `this._initialized` is supposed to be that countermeasure.

I agree that it smells. There should be a better place for the autoselection logic to run exactly once, right after `this.challenge` is loaded. What's that place? It's probably described somewhere in the `lit` documentation, but I haven't found it yet. Do you have any suggestions?

kensternberg-authentik replied:
Ah, okay, that makes a lot of sense.  It _is_ annoying that Lit's lifecycle has a `firstUpdated()` but not a `firstWillupdate()`, which is more or less what you're simulating here.

kensternberg-authentik reviewed (changes requested):
Change required: Do not mutate object state within the `render()` phase.

kensternberg-authentik committed:
web: move mutator block into the `willUpdate` override

Removed the section of code from the renderer that updates the state of the component;
Mutating in the middle of a render is strongly discouraged.  This block contains an
algorithm for determining if the selectedDeviceChallenge should be set and how; since
`selectedDeviceChallenge` is a state, we don't want to be changing it outside of those
lifecycle methods that do not trigger a rerender.

kensternberg-authentik commented on web/src/flow/stages/authenticator_validate/AuthenticatorValidateStage.ts:
```diff
@@ -112,6 +115,42 @@ export class AuthenticatorValidateStage
         `);
     }
 
+    willUpdate(_changed: PropertyValues<this>) {
```
This code block chooses which `selectedDeviceChallenge` to pick, if any.  I've moved it here because `willUpdate` is a special method; changes to states and properties within it will not trigger a re-draw, so this can be set without causing Lit to draw this component twice.  I also put in early returns to show that it's a selective, progressive algorithm; if one succeeds, a later one doesn't need to be examined.

kensternberg-authentik committed:
web: move styles() to top of class, extract custom CSS to a named block.

kensternberg-authentik commented on web/src/flow/stages/authenticator_validate/AuthenticatorValidateStage.ts:
```diff
@@ -33,6 +64,10 @@ export class AuthenticatorValidateStage
     >
     implements StageHost
 {
+    static get styles(): CSSResult[] {
```
I extracted the customCSS because it was big enough it distracted from the main purpose of the class, so this makes the entirety a bit more readable.

Also, there's an order to Lit components that we've adopted: Styles, Properties, States, other class fields, constructor, connectedCallback, disconnectedCallback, event handlers, and then the lifecycle methods from `performUpdate()` through `render()` and finally `updated()`.  (Obviously, don't include any that you don't override!) Yes, this means that lifecycle methods that run after `render()` appear after your `render()` method in the class.  Helper methods (sub-renderers, formatters, etc) go _above_ the first lifecycle method in which they're used, in the order in which they appear in the lifecycle method.  We're not perfect about this, but we are getting better.

kensternberg-authentik committed:
lint: collapse multiple early returns, missing curly brace.

gergosimonyi committed:
Merge branch 'main' into authenticator_validate/remember-last-2fa-device

gergosimonyi committed:
autoselect device only once even if the user only has 1 device

BeryJu commented on authentik/stages/authenticator_validate/stage.py:
```diff
@@ -218,6 +218,13 @@ def get_device_challenges(self) -> list[dict]:
                     "device_uid": device.pk,
                     "challenge": get_challenge_for_device(self.request, stage, device),
                 }
+                | (
```
I think it'd be cleaner to have this in the same dict instead of using the `|` operator

BeryJu reviewed (commented):
One more general comment based on the comment i made on the linked issue:
```
Our eventual implementation for this would save the last used authenticator for a pending user, there was the idea of saving this as a cookie like the validation does, however that would only save this on a specific device and not centrally per-user. (Also feel free to open a PR to add this)
```

I think it might be useful to do this too and have that as a higher priority?  There might be a bit of an advantage to that having that too if users use a specific authenticator always on a specific browser? Tbh I'm not 100% sure

gergosimonyi committed:
make `DeviceChallenge.last_used` nullable instead of optional

gergosimonyi committed:
Merge branch 'main' into authenticator_validate/remember-last-2fa-device

BeryJu reviewed (changes requested):
Needs docs otherwise LGTM

gergosimonyi committed:
Merge branch 'main' into authenticator_validate/remember-last-2fa-device

gergosimonyi committed:
clarify button text

gergosimonyi committed:
fix typo

gergosimonyi committed:
add docs for automatic device selection

tanberry commented on website/docs/flow/stages/authenticator_validate/index.md:
```diff
@@ -87,3 +87,7 @@ Optionally restrict which WebAuthn device types can be used to authenticate.
+
+#### Automatic device selection
+
+If the user has more than one device, a device picker screen is shown. Once the user successfully authenticates with a device, that device is marked as "last used". In subsequent challenges, the last used device will be automatically selected for the user. (Should they wish to use another device, the user can return to the device picker screen.)
```
```suggestion
If the user has more than one device, the user is prompted to select which device they want to use for validation. After the user successfully authenticates with a certain device, that device is marked as "last used". In subsequent logins, the last used device will be automatically selected for the user. Should they wish to use another device, the user can return to the device selection screen.
```

tanberry reviewed (approved):
This is a nice  feature! And nice docs. ;-) A few edits, then LGTM!

gergosimonyi committed:
update docs

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Simonyi Gergő <28359278+gergosimonyi@users.noreply.github.com>

gergosimonyi committed:
Merge branch 'main' into authenticator_validate/remember-last-2fa-device

gergosimonyi committed:
fix punctuation
